### PR TITLE
conv2d (Fixes #2785)

### LINKED ISSUE #2785: conv2d

### PR DESCRIPTION
**Pull request by Montassar-bdh**
Hi,
I reformatted conv2d layer (#2785). I couldn't work on the docstring examples because I didn't know from where to start.

### CODE REVIEW

**Comment by VedPatwardhan** [comment]
Hi @Montassar-bdh, please add the docstring examples to the functions referring to the [guide](https://lets-unify.ai/ivy/deep_dive/11_docstring_examples.html) in the docs.

**Review by Momentum-TN — COMMENTED** [review]
format array and container layer files

**Comment by VedPatwardhan** [comment]
Hi @Montassar-bdh, the docstring examples are yet to be added, feel free to request my review through the right pane when it's done. Thanks 🙂

**Comment by Montassar-bdh** [comment]
Hi @VedPatwardhan, I would like to have your review to the PR. I made a mistake by entring the review mode using my other github account and I can't remove it from the reviewer right pane and assign you instead.

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @Montassar-bdh, please resolve the merge conflicts in your PR so we can see if the reformatting-related tests pass for your function. Thanks 🙂

**Comment by Montassar-bdh** [comment]
@VedPatwardhan I have fixed the conflict. Could you please help me solving the unsuccessful checks?

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @Montassar-bdh, please update the way you are passing parameters inside all functions, basically all arguments before `/` in the parameter list are positional-only arguments and all those after `*` are keyword-only arguments. This is causing the docstring tests to fail. Also, the same applies to the way you are calling functions inside these functions too, for e.g. the way you are calling `ivy.conv2d` inside the `conv2d` instance and static methods, etc. should also follow the same format. Further, there are some changes to be made to the functional version of `conv2d` too inside `ivy/functional/ivy/layers.py` as it's docstring is also not in the format specified in the docs.
Feel free to re-request my review once you're done with the changes.
Let me know if you have any questions. Thanks 🙂

**Comment by Montassar-bdh** [comment]
@VedPatwardhan Thank you for the clarification. I will update them and re-request your review when I complete the changes 🙂

**Comment by Montassar-bdh** [comment]
Hey @VedPatwardhan, I checked the [`array docs`](https://lets-unify.ai/ivy/data_classes/array.html) and [`functional layers docs`](https://lets-unify.ai/ivy/functional/ivy/layers.html#ivy.functional.ivy.layers.conv2d). The function signature in the code is identical to the docs.

In container [container data class docs](https://lets-unify.ai/ivy/data_classes/container.html#ivy.container.container.Container.conv2d) I am not sure whether I should put the same signature (`*,key_chains=None, to_apply=True, prune_unapplied=False, map_sequences=False, out: Optional[[Container](https://lets-unify.ai/ivy/data_classes/container/container/Container.html#ivy.container.container.Container)] = None, **kwargs`)

**Comment by Montassar-bdh** [comment]
I made a merging with the master branch. The number of successful tests decreased

**Comment by VedPatwardhan** [comment]
Hi @Montassar-bdh, that's okay, we just need to confirm that the changes made by you don't cause failures. There seem to be some errors in the docstring tests for your function, related to wrongly passing arguments. But firstly, you should update the container static and instance methods of `conv2d` with additional arguments like `key_chains`, etc. according to the way they're added in the examples specified in the docs. 
Also, please re-request my review through the right pane once you're done with the changes, otherwise it gets difficult for me to know if you're done with the changes or not.
Thanks 🙂

**Comment by VedPatwardhan** [comment]
Hi @Montassar-bdh, are you still working on this PR, please message in the chats if you are, because otherwise this PR may get closed automatically due to inactivity. Thanks 🙂

**Comment by Montassar-bdh** [comment]
Hey @VedPatwardhan, I am still working on the PR. Thank you for letting me know.

**Comment by Montassar-bdh** [comment]
About those parameters that I told you about, they were related to the `Container` data class, not with `ContainerWithLayers` base class.

I faced error in lint check caused by a file that I didn't touch! (it was successful before my last commit)
For the docstring test, I don't understand why it's failing even though I followed the docstring example of other functions (like `conv1d`)

I would like to kindly request your help.

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @Montassar-bdh
1. You don't need to worry about the failures not related to your PRs, there are some failures on master which may get reflected in PRs, so as long as the tests for your function are passing, we can merge your PR.
2. The docstring tests are failing for your function because it seems like you've passed some positional-only arguments as keyword arguments in the examples or the other way round. In order to fix it, please checkout the function signature of the `conv2d` function and make function calls accordingly into the examples. 
3. Also, as I mentioned in one of the previous reviews, the docstring for the `conv2d` function in `ivy/functional/ivy/layers.py` also comes under your task, so please check for it's correctness as well. I think you may find the reason for the docstring failure in that file.
4. Also, please update your branch with the latest master code.
5. One final suggestion, please refer the examples provided in the docs rather than from any other function as those may also have been added by some other applicants, which raises the possibility of an error.

Let me know if you have any questions. Thanks 🙂

**Comment by VedPatwardhan** [comment]
Hi @Montassar-bdh, please message in the chats if you're still working on this PR, because it may get closed automatically in a few days due to inactivity. Thanks 🙂

**Comment by Montassar-bdh** [comment]
@VedPatwardhan  Thank you for the reminder. Yes I am still working on the issue and I would like to have your opinion on the following issue that I got when testing the docstring.
Which function is causing the positional argument error: conv2d or view ?
For JAX:
```
The following methods had failing docstrings:
Container.conv2d
Array.conv2d
-------------------------------------------------------------
output for  conv2d  on run:  _view()takesfrom1to3positionalargumentsbut5weregivenjaxx=ivy.Container(a=ivy.eye(3,3).view(1,3,3,1),b=ivy.eye(5,5).view(1,5,5,1)),'DeviceArray'objecthasnoattribute'__getattr__'jaxfilters=ivy.array([[2,0,1],[1,3,1],[0,1,1]]).unsqueeze(-1).unsqueeze(-1).float(),name'x'isnotdefinedjaxresult=x.conv2d(filters,(2,),'SAME'),name'result'isnotdefinedjaxprint(result) 
output in docs : {a:ivy.array([[[[4.],[0.]],[[1.],[5.]]]]),b:ivy.array([[[[4.],[0.],[0.]],[[1.],[6.],[0.]],[[0.],[1.],[5.]]]])} 
 _viewtakesfrom1to3positionalargumentsbut5weregivenjaxx=ivy.Containera=ivy.eye3  !=  4. 

output for  conv2d  on run:  conv_general_dilatedwindowandwindow_stridesmusthavethesamenumberofdimensions,butgot2and1jaxresult=x.conv2d(filters,(1,),'SAME',data_format='NHWC',dilations=(1,)),name'result'isnotdefinedjaxprint(result) 
output in docs : ivy.array([[[[2.],[4.],[6.]],[[3.],[6.],[9.]],[[2.],[4.],[6.]]]]) 
 conv_general_dilatedwindowandwindow_stridesmusthavethesamenumberofdimensions  !=  2. 
```

For numpy:
```
Container.conv2d
------------------------------------------------------------- Captured stdout call -------------------------------------------------------------
output for  conv2d  on run:  view()takesfrom0to2positionalargumentsbut4weregivennumpyx=ivy.Container(a=ivy.eye(3,3).view(1,3,3,1),b=ivy.eye(5,5).view(1,5,5,1)),'numpy.ndarray'objecthasnoattribute'__getattr__'numpyfilters=ivy.array([[2,0,1],[1,3,1],[0,1,1]]).unsqueeze(-1).unsqueeze(-1).float(),name'x'isnotdefinednumpyresult=x.conv2d(filters,(2,),'SAME'),name'result'isnotdefinednumpyprint(result) 
output in docs : {a:ivy.array([[[[4.],[0.]],[[1.],[5.]]]]),b:ivy.array([[[[4.],[0.],[0.]],[[1.],[6.],[0.]],[[0.],[1.],[5.]]]])} 
 viewtakesfrom0to2positionalargumentsbut4weregivennumpyx=ivy.Containera=ivy.eye3  !=  4. 
```
 
And for tensorflow:
```
output for  conv2d  on run:  'tensorflow.python.framework.ops.EagerTensor'objecthasnoattribute'view'tensorflowx=ivy.Container(a=ivy.eye(3,3).view(1,3,3,1),b=ivy.eye(5,5).view(1,5,5,1)),'tensorflow.python.framework.ops.EagerTensor'objecthasnoattribute'unsqueeze'tensorflowfilters=ivy.array([[2,0,1],[1,3,1],[0,1,1]]).unsqueeze(-1).unsqueeze(-1).float(),name'x'isnotdefinedtensorflowresult=x.conv2d(filters,(2,),'SAME'),name'result'isnotdefinedtensorflowprint(result) 
output in docs : {a:ivy.array([[[[4.],[0.]],[[1.],[5.]]]]),b:ivy.array([[[[4.],[0.],[0.]],[[1.],[6.],[0.]],[[0.],[1.],[5.]]]])} 
 'tensorflow.python.framework.ops.EagerTensor'objecthasnoattribute'view'tensorflowx=ivy.Containera=ivy.eye3  !=  4.
﻿```

**Comment by VedPatwardhan** [comment]
> @VedPatwardhan Thank you for the reminder. Yes I am still working on the issue and I would like to have your opinion on the following issue that I got when testing the docstring. Which function is causing the positional argument error: conv2d or view ? For JAX:
> 
> ```
> The following methods had failing docstrings:
> Container.conv2d
> Array.conv2d
> -------------------------------------------------------------
> output for  conv2d  on run:  _view()takesfrom1to3positionalargumentsbut5weregivenjaxx=ivy.Container(a=ivy.eye(3,3).view(1,3,3,1),b=ivy.eye(5,5).view(1,5,5,1)),'DeviceArray'objecthasnoattribute'__getattr__'jaxfilters=ivy.array([[2,0,1],[1,3,1],[0,1,1]]).unsqueeze(-1).unsqueeze(-1).float(),name'x'isnotdefinedjaxresult=x.conv2d(filters,(2,),'SAME'),name'result'isnotdefinedjaxprint(result) 
> output in docs : {a:ivy.array([[[[4.],[0.]],[[1.],[5.]]]]),b:ivy.array([[[[4.],[0.],[0.]],[[1.],[6.],[0.]],[[0.],[1.],[5.]]]])} 
>  _viewtakesfrom1to3positionalargumentsbut5weregivenjaxx=ivy.Containera=ivy.eye3  !=  4. 
> 
> output for  conv2d  on run:  conv_general_dilatedwindowandwindow_stridesmusthavethesamenumberofdimensions,butgot2and1jaxresult=x.conv2d(filters,(1,),'SAME',data_format='NHWC',dilations=(1,)),name'result'isnotdefinedjaxprint(result) 
> output in docs : ivy.array([[[[2.],[4.],[6.]],[[3.],[6.],[9.]],[[2.],[4.],[6.]]]]) 
>  conv_general_dilatedwindowandwindow_stridesmusthavethesamenumberofdimensions  !=  2. 
> ```
> 
> For numpy:
> 
> ```
> Container.conv2d
> ------------------------------------------------------------- Captured stdout call -------------------------------------------------------------
> output for  conv2d  on run:  view()takesfrom0to2positionalargumentsbut4weregivennumpyx=ivy.Container(a=ivy.eye(3,3).view(1,3,3,1),b=ivy.eye(5,5).view(1,5,5,1)),'numpy.ndarray'objecthasnoattribute'__getattr__'numpyfilters=ivy.array([[2,0,1],[1,3,1],[0,1,1]]).unsqueeze(-1).unsqueeze(-1).float(),name'x'isnotdefinednumpyresult=x.conv2d(filters,(2,),'SAME'),name'result'isnotdefinednumpyprint(result) 
> output in docs : {a:ivy.array([[[[4.],[0.]],[[1.],[5.]]]]),b:ivy.array([[[[4.],[0.],[0.]],[[1.],[6.],[0.]],[[0.],[1.],[5.]]]])} 
>  viewtakesfrom0to2positionalargumentsbut4weregivennumpyx=ivy.Containera=ivy.eye3  !=  4. 
> ```
> 
> And for tensorflow:
> 
> ```
> output for  conv2d  on run:  'tensorflow.python.framework.ops.EagerTensor'objecthasnoattribute'view'tensorflowx=ivy.Container(a=ivy.eye(3,3).view(1,3,3,1),b=ivy.eye(5,5).view(1,5,5,1)),'tensorflow.python.framework.ops.EagerTensor'objecthasnoattribute'unsqueeze'tensorflowfilters=ivy.array([[2,0,1],[1,3,1],[0,1,1]]).unsqueeze(-1).unsqueeze(-1).float(),name'x'isnotdefinedtensorflowresult=x.conv2d(filters,(2,),'SAME'),name'result'isnotdefinedtensorflowprint(result) 
> output in docs : {a:ivy.array([[[[4.],[0.]],[[1.],[5.]]]]),b:ivy.array([[[[4.],[0.],[0.]],[[1.],[6.],[0.]],[[0.],[1.],[5.]]]])} 
>  'tensorflow.python.framework.ops.EagerTensor'objecthasnoattribute'view'tensorflowx=ivy.Containera=ivy.eye3  !=  4.
> ﻿```
> ```

Hi @Montassar-bdh, you're task is related to the `conv2d` function which may or may not is related to this one failure. But as I've said in the previous review, please pull the latest code from master on your branch as the code on master keeps getting updated frequently. Feel free to also look into the other things I've mentioned in my previous review and re-request my review once you're done with them or have any questions regarding them. Thanks 🙂

**Comment by Montassar-bdh** [comment]
Hi again!
Thank you for all the hints and tips.
I went through the previous reviews and the docs. I made the necessary changes. Now I got the following output for the docstring test and I would like to have your review:

JAX:
``` 
E             Container.conv2d
E             Array.conv2d
-----------------------------------------
 output for  conv2d  on run:  _view()takesfrom1to3positionalargumentsbut5weregivenjaxx=ivy.Container(a=ivy.eye(3,3).view(1,3,3,1),b=ivy.eye(5,5).view(1,5,5,1)),'DeviceArray'objecthasnoattribute'__getattr__'jaxfilters=ivy.array([[2,0,1],[1,3,1],[0,1,1]]).unsqueeze(-1).unsqueeze(-1).float(),name'x'isnotdefinedjaxresult=x.conv2d(filters,(2,),'SAME'),name'result'isnotdefinedjaxprint(result) 
output in docs : {a:ivy.array([[[[4.],[0.]],[[1.],[5.]]]]),b:ivy.array([[[[4.],[0.],[0.]],[[1.],[6.],[0.]],[[0.],[1.],[5.]]]])} 
 _viewtakesfrom1to3positionalargumentsbut5weregivenjaxx=ivy.Containera=ivy.eye3  !=  4. 

 output for  conv2d  on run:  conv_general_dilatedwindowandwindow_stridesmusthavethesamenumberofdimensions,butgot2and1jaxresult=x.conv2d(filters,(1,),'SAME',data_format='NHWC',dilations=(1,)),name'result'isnotdefinedjaxprint(result) 
output in docs : ivy.array([[[[2.],[4.],[6.]],[[3.],[6.],[9.]],[[2.],[4.],[6.]]]]) 
 conv_general_dilatedwindowandwindow_stridesmusthavethesamenumberofdimensions  !=  2. 
```

Numpy:
``` 
E             Container.conv2d
-----------
output for  conv2d  on run:  view()takesfrom0to2positionalargumentsbut4weregivennumpyx=ivy.Container(a=ivy.eye(3,3).view(1,3,3,1),b=ivy.eye(5,5).view(1,5,5,1)),'numpy.ndarray'objecthasnoattribute'__getattr__'numpyfilters=ivy.array([[2,0,1],[1,3,1],[0,1,1]]).unsqueeze(-1).unsqueeze(-1).float(),name'x'isnotdefinednumpyresult=x.conv2d(filters,(2,),'SAME'),name'result'isnotdefinednumpyprint(result) 
output in docs : {a:ivy.array([[[[4.],[0.]],[[1.],[5.]]]]),b:ivy.array([[[[4.],[0.],[0.]],[[1.],[6.],[0.]],[[0.],[1.],[5.]]]])} 
 viewtakesfrom0to2positionalargumentsbut4weregivennumpyx=ivy.Containera=ivy.eye3  !=  4. 
```


Tensorflow:
```
E             Container.conv2d
--------------------------------------
output for  conv2d  on run:  'tensorflow.python.framework.ops.EagerTensor'objecthasnoattribute'view'tensorflowx=ivy.Container(a=ivy.eye(3,3).view(1,3,3,1),b=ivy.eye(5,5).view(1,5,5,1)),'tensorflow.python.framework.ops.EagerTensor'objecthasnoattribute'unsqueeze'tensorflowfilters=ivy.array([[2,0,1],[1,3,1],[0,1,1]]).unsqueeze(-1).unsqueeze(-1).float(),name'x'isnotdefinedtensorflowresult=x.conv2d(filters,(2,),'SAME'),name'result'isnotdefinedtensorflowprint(result) 
output in docs : {a:ivy.array([[[[4.],[0.]],[[1.],[5.]]]]),b:ivy.array([[[[4.],[0.],[0.]],[[1.],[6.],[0.]],[[0.],[1.],[5.]]]])} 
 'tensorflow.python.framework.ops.EagerTensor'objecthasnoattribute'view'tensorflowx=ivy.Containera=ivy.eye3  !=  4. 
```

Torch

```
E             Container.conv2d
------------------------------------
output for  conv2d  on run:  numberofdimsdon'tmatchinpermutetorchresult=x.conv2d(filters,(2,),'SAME'),name'result'isnotdefinedtorchprint(result) 
output in docs : {a:ivy.array([[[[4.],[0.]],[[1.],[5.]]]]),b:ivy.array([[[[4.],[0.],[0.]],[[1.],[6.],[0.]],[[0.],[1.],[5.]]]])} 
 numberofdimsdon'tmatchinpermutetorchresult=x.conv2dfilters  !=  4.
```

Thanks 😄

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @Montassar-bdh, thanks for making the changes as suggested. I'm not sure where you found those logs, but the logs of the docstring tests fail with this message on the [CI](https://github.com/unifyai/ivy/runs/8169277026?check_suite_focus=true)
```
output for  conv2d  on run:  numberofdimsdon'tmatchinpermutetorchresult=x.conv2d(filters,(2,),'SAME'),name'result'isnotdefinedtorchprint(result) 
output in docs : {a:ivy.array([[[[4.],[0.]],[[1.],[5.]]]]),b:ivy.array([[[[4.],[0.],[0.]],[[1.],[6.],[0.]],[[0.],[1.],[5.]]]])} 
 numberofdimsdon'tmatchinpermutetorchresult=x.conv2dfilters  !=  4. 
```
Please check through the Container examples for fixing this failure. Also, as I've mentioned in a number of previous reviews, the docstring for the `conv2d` function inside **`ivy/functional/ivy/layers.py** is also not according to the format specified in the [docs](https://lets-unify.ai/ivy/deep_dive/11_docstring_examples.html). 
**Please update that docstring too, as a part of the [function formatting task](https://lets-unify.ai/ivy/contributing/4_open_tasks.html#function-formatting)**

Feel free to re-request my review once you're done with the changes, rest looks good to me!
Thanks 🙂

**Comment by Montassar-bdh** [comment]
Hi @VedPatwardhan,

The logs are in the same CI that you mentioned. I understand that the docstring test evaluates for all the backends (JAX, numpy, tensorflow and pytorch). If you go [this line](https://github.com/unifyai/ivy/runs/8169277026?check_suite_focus=true#step:4:60), you will find the test for JAX and the results related to `conv2d` are here in [the following line](https://github.com/unifyai/ivy/runs/8169277026?check_suite_focus=true#step:4:214).


About this message below,
> ```
> output for  conv2d  on run:  numberofdimsdon'tmatchinpermutetorchresult=x.conv2d(filters,(2,),'SAME'),name'result'isnotdefinedtorchprint(result) 
> output in docs : {a:ivy.array([[[[4.],[0.]],[[1.],[5.]]]]),b:ivy.array([[[[4.],[0.],[0.]],[[1.],[6.],[0.]],[[0.],[1.],[5.]]]])} 
>  numberofdimsdon'tmatchinpermutetorchresult=x.conv2dfilters  !=  4. 
> ```

This failure is raised by torch backend. I have fixed the issue.



> Also, as I've mentioned in a number of previous reviews, the docstring for the `conv2d` function inside **`ivy/functional/ivy/layers.py** is also not according to the format specified in the [docs](https://lets-unify.ai/ivy/deep_dive/11_docstring_examples.html).
> **Please update that docstring too, as a part of the [function formatting task](https://lets-unify.ai/ivy/contributing/4_open_tasks.html#function-formatting)**

This is the functional conv2d docstring:
```
    """Computes a 2-D convolution given 4-D input x and filters arrays.

    Parameters
    ----------
    x
        Input image *[batch_size,h,w,d_in]*.
    filters
        Convolution filters *[fh,fw,d_in,d_out]*.
    strides
        The stride of the sliding window for each dimension of input.
    padding
        SAME" or "VALID" indicating the algorithm, or list indicating the per-dimension
        paddings.
    data_format
        NHWC" or "NCHW". Defaults to "NHWC".
    dilations
        The dilation factor for each dimension of input. (Default value = 1)
    out
        optional output array, for writing the result to. It must have a shape that the
        inputs broadcast to.

    Returns
    -------
    ret
        The result of the convolution operation.

    Both the description and the type hints above assumes an array input for simplicity,
    but this function is *nestable*, and therefore also accepts :code:`ivy.Container`
    instances in place of any of the arguments.


    Functional Examples
    -------------------

    With :code:`ivy.Array` input:

    >>> x = ivy.array([[[[1.], [2.0],[3.]], \
                      [[1.], [2.0],[3.]], \
                      [[1.], [2.0],[3.]]]]) #NHWC

    >>> filters = ivy.array([[[[0.]],[[1.]],[[0.]]], \
                             [[[0.]],[[1.]], [[0.]]], \
                             [[[0.]],[[1.]], [[0.]]]]) #HWIO
    >>> result = ivy.conv2d(x, filters, (1,), 'SAME', data_format='NHWC',\
     dilations= (1,))
    >>> print(result)
    ivy.array([[
              [[2.],[4.],[6.]],
              [[3.],[6.],[9.]],
              [[2.],[4.],[6.]]
              ]])

    With :code:`ivy.NativeArray` input:

    >>> x = ivy.native_array(ivy.random_normal(mean=0, std=1, shape=[1, 32, 32, 3]))
    >>> filters = ivy.native_array(ivy.random_normal(mean=0, std=1, \
                                   shape=[3, 5, 3, 5])) #HWIO
    >>> result = ivy.conv2d(x, filters, [2, 1], 'VALID') \
        #non-square filter with unequal stride and valid padding
    >>> print(result.shape)
    (1, 15, 28, 5)


    With a mix of :code:`ivy.Array` and :code:`ivy.Container` inputs:

    >>> x = ivy.Container(a = ivy.eye(3, 3).view(1, 3, 3, 1), \
                          b = ivy.eye(5, 5).view(1, 5, 5, 1))
    >>> filters = ivy.array([[2, 0, 1], \
                             [1, 3, 1], \
                             [0, 1, 1]]).unsqueeze(-1).unsqueeze(-1).float()
    >>> result = ivy.conv2d(x, filters, (2,), 'SAME')
    >>> print(result)
    {
        a:ivy.array([[[[4.],[0.]],[[1.],[5.]]]]),
        b:ivy.array([[[[4.],[0.],[0.]],[[1.],[6.],[0.]],[[0.],[1.],[5.]]]])
    }

    """
```

The examples are following to what is mentioned in the docs. I have re-read it and didn't understand what I should change. Could you please give me a hint ?
Thanks

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @Montassar-bdh, thanks for making the changes. The docstring tests have started passing for your function. About the functional docstring there are a number of changes to be made,
1. There shouldn't be examples with the header `With ivy.NativeArray input`.
2. `Functional Examples` should be renamed to `Examples`
3. The examples should also include containers under the headers `With one ivy.Container input` and `With multiple ivy.Container inputs`.

Feel free to re-request my review once you're done with the changes, I think we could merge your PR as everything else seems complete. 
Let me know if you have any questions. Thanks 🙂

**Comment by Montassar-bdh** [comment]
Hi @VedPatwardhan,
I hope you are well.
Thank you for the clarification. 

>The examples should also include containers under the headers `With one ivy.Container input` and `With multiple ivy.Container inputs`.

I added the templates of the examples. Please tell me if I am on the right path.

Thanks 😄

File: ivy/functional/ivy/layers.py [hunk 1]
```diff
@@ -676,7 +676,7 @@ def conv2d(
     instances in place of any of the arguments.
 
 
-    Functional Examples
+    Examples
     -------------------
```
Comment (VedPatwardhan) [inline ivy/functional/ivy/layers.py:680]: The number of `-` should be equal to the number of characters in `Examples`

File: ivy/functional/ivy/layers.py [hunk 2]
```diff
@@ -676,7 +676,7 @@ def conv2d(
...
```
Comment (VedPatwardhan) [inline ivy/functional/ivy/layers.py:681]: There shouldn't be a blank line here

File: ivy/functional/ivy/layers.py [hunk 3]
```diff
@@ -697,16 +697,27 @@ def conv2d(
...
-
-    >>> x = ivy.native_array(ivy.random_normal(mean=0, std=1, shape=[1, 32, 32, 3]))
-    >>> filters = ivy.native_array(ivy.random_normal(mean=0, std=1, \
-                                   shape=[3, 5, 3, 5])) #HWIO
-    >>> result = ivy.conv2d(x, filters, [2, 1], 'VALID') \
-        #non-square filter with unequal stride and valid padding
-    >>> print(result.shape)
-    (1, 15, 28, 5)
+    With one :code:`ivy.Container` input:
+    >>> x = ivy.Container(a=ivy.)
```
Comment (VedPatwardhan) [inline ivy/functional/ivy/layers.py:701]: Please add a blank line here

Comment (VedPatwardhan) [reply ivy/functional/ivy/layers.py:701]: Invalid syntax on this line

File: ivy/functional/ivy/layers.py [hunk 4]
```diff
@@ -697,16 +697,27 @@ def conv2d(
...
+    >>> filters = ivy.
```
Comment (VedPatwardhan) [inline ivy/functional/ivy/layers.py:702]: Invalid syntax

File: ivy/functional/ivy/layers.py [hunk 5]
```diff
@@ -697,16 +697,27 @@ def conv2d(
...
+    >>> result = ivy.conv2d(x, filters, (2,), 'SAME')
+    >>> print(result)
+    {
+        a:ivy.array([])
+    }
 
+    With multiple :code:`ivy.Container` inputs:
+    >>> x = ivy.Container(a = ivy.,  \
```
Comment (VedPatwardhan) [inline ivy/functional/ivy/layers.py:710]: Invalid syntax, you should be creating an array using `ivy.array` instead of just `ivy.`

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @Montassar-bdh, thanks for making the changes. I have requested a few more changes to get the docstring tests and numpy style docstring tests passing for your function, feel free to make those changes and re-request my review once you're done. Thanks 🙂

**Comment by Montassar-bdh** [comment]
Hi @VedPatwardhan,
Thank you for the comments.
About the syntax errors, I just created the templates for the examples to check if we are on the same page.
I will fix them. By the way, how can I create my own examples using custom inputs? importing ivy and working with the functions raise errors.

**Comment by VedPatwardhan** [comment]
Hi @Montassar-bdh, for creating examples, you could just create an example.py file inside the root directory of the project. Then you'll have to import ivy, set a backend and run an example. Do let me know if you run into any issues while doing that. Thanks 🙂

**Comment by Montassar-bdh** [comment]
Hi @VedPatwardhan,
I completed the docstring examples as requested.
Thanks 😄

File: ivy/array/layers.py [hunk 6]
```diff
@@ -232,7 +232,75 @@ def depthwise_conv2d(
...
+        >>> print(result)
+        ivy.array([[
+                  [[2.],[4.],[6.]],
+                  [[3.],[6.],[9.]],
+                  [[2.],[4.],[6.]]
+                  ]])
+
+        """
+        return ivy.conv2d(
+            self._data,
```
Comment (VedPatwardhan) [inline ivy/array/layers.py:295]: Please pass `self` instead of `self._data`

File: ivy/container/layers.py [hunk 7]
```diff
@@ -408,6 +408,49 @@ def static_conv2d(
...
+            that the inputs broadcast to.
+        
+        Returns
+        -------
+        ret
+            The result of the convolution operation.
+        
+        Examples
+        --------
+        >>> x = ivy.Container(a = ivy.eye(3, 3).view(1, 3, 3, 1), \
```
Comment (VedPatwardhan) [inline ivy/container/layers.py:442]: `view` isn't a function in Ivy's API, I'd suggest you to use functions present in ivy and also test your examples to see if the results are as expected

File: ivy/container/layers.py [hunk 8]
```diff
@@ -408,6 +408,49 @@ def static_conv2d(
...
+                              b = ivy.eye(5, 5).view(1, 5, 5, 1))
+        >>> filters = ivy.array([[2, 0, 1], \
+                                [1, 3, 1], \
+                                [0, 1, 1]]).unsqueeze(-1).unsqueeze(-1).float()
```
Comment (VedPatwardhan) [inline ivy/container/layers.py:446]: `unsqueeze` and `float` aren't functions in the Ivy API either, please test your examples to see if the results are as expected.

File: ivy/container/layers.py [hunk 9]
```diff
@@ -438,7 +481,50 @@ def conv2d(
...
+            that the inputs broadcast to.
+        
+        Returns
+        -------
+        ret
+            The result of the convolution operation.
+        
+        Examples
+        --------
+        >>> x = ivy.Container(a = ivy.eye(3, 3).view(1, 3, 3, 1), \
```
Comment (VedPatwardhan) [inline ivy/container/layers.py:515]: Same problem as the container instance method `conv2d`

File: ivy/container/layers.py [hunk 10]
```diff
@@ -438,7 +481,50 @@ def conv2d(
...
+                              b = ivy.eye(5, 5).view(1, 5, 5, 1))
+        >>> filters = ivy.array([[2, 0, 1], \
```
Comment (VedPatwardhan) [inline ivy/container/layers.py:517]: Same problem as the container instance method `conv2d`

File: ivy/functional/ivy/layers.py [hunk 11]
```diff
@@ -697,16 +696,36 @@ def conv2d(
...
-    >>> result = ivy.conv2d(x, filters, [2, 1], 'VALID') \
-        #non-square filter with unequal stride and valid padding
-    >>> print(result.shape)
-    (1, 15, 28, 5)
+    With one :code:`ivy.Container` input:
+    
+    >>> x = ivy.Container(a=ivy.array([[[[1.], [2.0],[3.]], \
+                                        [[1.], [2.0],[3.]], \
+                                        [[1.], [2.0],[3.]]]]))
+    >>> filters = ivy.eye(3, 3).unsqueeze(-1).unsqueeze(-1).float()
```
Comment (VedPatwardhan) [inline ivy/functional/ivy/layers.py:704]: `unsqueeze` and `float` aren't functions in the Ivy API either, please test your examples to see if the results are as expected.

File: ivy/functional/ivy/layers.py [hunk 12]
```diff
@@ -697,16 +696,36 @@ def conv2d(
...
+    >>> result = ivy.conv2d(x, filters, (2,), 'SAME', data_format='NHWC', \
+        dilations= (1,))
+    >>> print(result)
+    {
+        a:ivy.array([[[[3.], [3.]], [[1.], [5.]]]])
+    }
 
+    With multiple :code:`ivy.Container` inputs:
+    >>> x = ivy.Container(a = (3, 3).view(1, 3, 3, 1),  \
```
Comment (VedPatwardhan) [inline ivy/functional/ivy/layers.py:713]: `view` isn't a function in Ivy's API, I'd suggest you to use functions present in ivy and also test your examples to see if the results are as expected

File: ivy/functional/ivy/layers.py [hunk 13]
```diff
@@ -697,16 +696,36 @@ def conv2d(
...
+                          b = (4, 4).view(1, 4, 4, 1), \
+                          c = (5, 5).view(1, 5, 5, 1))
+    >>> filters = ivy.array([[1, 1, 1], \
```
Comment (VedPatwardhan) [inline ivy/functional/ivy/layers.py:716]: `unsqueeze` and `float` aren't functions in the Ivy API either, please test your examples to see if the results are as expected.

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @Montassar-bdh, I've suggested a few more changes to the PR, mainly regarding the use of functions present in Ivy's API in the docstring examples. I'd also suggest you to test the examples like I mentioned previously as that'll make it easier for you to fix the issues.
Feel free to re-request my review once you're done. Thanks 🙂

**Comment by Montassar-bdh** [comment]
Hi @VedPatwardhan,
I have reformatted docstring examples using only ivy API functions as requested

I checked the dedicated Github action and found that only JAX framework is raising errors related to conv2d. ([Container.conv2d](https://github.com/unifyai/ivy/actions/runs/3083894084/jobs/4985390098#step:4:244) and [Array.conv2d](https://github.com/unifyai/ivy/actions/runs/3083894084/jobs/4985390098#step:4:292))

Everything else is looking good to me.

Looking forward to your review 😄

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @Montassar-bdh, thanks for making the changes! The docstring tests seem to be failing because of an issue with the dimensions of inputs passed to the [array instance method and the container instance method](https://github.com/unifyai/ivy/actions/runs/3084211146/jobs/4986097581#step:4:188), please test those examples and make changes accordingly.

**Comment by Montassar-bdh** [comment]
Hi @VedPatwardhan ,

Thank you for the guidance and the support. I have fixed the examples. All the issues related to `conv2d` reformatting has been addressed successfully. There is no issues now.

**Review by VedPatwardhan — APPROVED** [review]
lgtm! Thanks for contributing, I'll merge your PR now 🙂

**Comment by Montassar-bdh** [comment]
Thank you for your guidance during this task.
I had great fun

**Pull request merged**